controllers/baseboardmanagement_controller.go

```diff
@@ -18,42 +18,233 @@ package controllers
+	return &BaseboardManagementReconciler{
+		client:           client,
+		scheme:           scheme,
+		bmcClientFactory: bmcClientFactory,
+		logger:           logger,
+	}
 }
 
+// baseboardManagementFieldReconciler defines a function to reconcile BaseboardManagement spec field
+type baseboardManagementFieldReconciler func(context.Context, *bmcv1alpha1.BaseboardManagement, BMCClient) error
+
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=baseboardmanagements/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BaseboardManagement object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile ensures the state of a BaseboardManagement.
+// Gets the BaseboardManagement object and uses the SecretReference to initialize a BMC Client.
+// Ensures the BMC power is set to the desired state.
+// Updates the status and conditions accordingly.
 func (r *BaseboardManagementReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BaseboardManagement", req.NamespacedName)
+	logger.Info("Reconciling BaseboardManagement")
+
+	// Fetch the BaseboardManagement object
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.client.Get(ctx, req.NamespacedName, baseboardManagement)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BaseboardManagement")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !baseboardManagement.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, baseboardManagement, logger)
+}
+
+func (r *BaseboardManagementReconciler) reconcile(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement, logger logr.Logger) (ctrl.Result, error) {
+	// Fetching username, password from SecretReference
+	// Requeue if error fetching secret
+	username, password, err := r.resolveAuthSecretRef(ctx, bm.Spec.Connection.AuthSecretRef)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BaseboardManagement %s/%s SecretReference: %v", bm.Namespace, bm.Name, err)
+	}
+
+	// TODO (pokearu): Remove port hardcoding
+	// Initializing BMC Client
+	bmcClient, err := r.bmcClientFactory(ctx, bm.Spec.Connection.Host, "623", username, password)
```

Comment:
Are you going to remove this hardcoding in this PR?

Reply:
I was planning to do a quick follow up PR. Was going to add it to the `BaseboardManagement` api connection struct.